### ISSUE TITLE
Overwriting and deleting files in mounted collections is broken in iRODS 4.2.8

### ISSUE THREAD

**Issue by mcoonen** [irods/irods#5457]
- [x] master
- [x] 4-2-stable
---------

### iRODS Version, OS and Version
iRODS 4.2.8 on Ubuntu 16.04

### What did you try to do?
Use `msiDataObjWrite` (with forceFlag enabled) to overwrite a file stored in an iRODS mounted collection. The code block below contains a rule to reproduce the problem.

Also, deletion of the file with `irm -f` is not working.

```
# Call with
#
# irule -r irods_rule_engine_plugin-irods_rule_language-instance -F overwriteFileInMountedColl.r "*irodsColl='/nlmumc/home/rods/mounted_coll'" "*irodsResc='rootResc'" "*phyDir='/tmp/mounted-coll'"

irule_dummy() {
    IRULE_overwriteFileInMountedColl(*irodsColl, *irodsResc, *phyDir);
}

IRULE_overwriteFileInMountedColl(*irodsColl, *irodsResc, *phyDir) {
    # Create the mounted collection
    msiCollCreate(*irodsColl, 0, *status)
    msiPhyPathReg(*irodsColl, *irodsResc, *phyDir, "mountPoint", *status);

    # Write data to file
    *Obj1 = "*irodsColl/foo1"
    *OFlags1 = "destRescName=rootResc++++forceFlag="
    *R_BUF1 = "This is the content for the file that will be written to the mounted collection in iRODS. It is longer than 100 bytes.\n"

    msiDataObjCreate(*Obj1,*OFlags1,*D_FD1);
    msiDataObjWrite(*D_FD1,*R_BUF1,*W_LEN1);
    msiDataObjClose(*D_FD1,*Status1);

    writeLine("stdout","Wrote file *Obj1");

    # Overwrite data in file
    *Obj2 = *Obj1
    *OFlags2 = "destRescName=rootResc++++forceFlag="
    *R_BUF2 = "This is the CHANGED content for the file that will be written to the mounted collection in iRODS. It is definitely longer than 100 bytes.\n"

    msiDataObjCreate(*Obj2,*OFlags2,*D_FD2);
    msiDataObjWrite(*D_FD2,*R_BUF2,*W_LEN2);
    msiDataObjClose(*D_FD2,*Status2);

    writeLine("stdout","Overwrote file *Obj2");
}

INPUT *irodsColl="", *irodsResc="", *phyDir=""
OUTPUT ruleExecOut
```

### Expected behavior
The example rule above would mount a collection, write 119 bytes to the file `foo1`, then overwrite `foo1` with 138 bytes.
This functionality works in iRODS 4.2.6, as demonstrated below.

Call the rule 
```
irule -r irods_rule_engine_plugin-irods_rule_language-instance -F overwriteFileInMountedColl.r "*irodsColl='/nlmumc/home/rods/mounted_coll'" "*irodsResc='rootResc'" "*phyDir='/tmp/mounted-coll'"

# stdout:
Wrote file /nlmumc/home/rods/mounted_coll/foo1
Overwrote file /nlmumc/home/rods/mounted_coll/foo1
```

List files in iRODS
```
ils -L /nlmumc/home/rods/mounted_coll

# stdout:
/nlmumc/home/rods/mounted_coll:
  rods         mountP rootResc;demoResc          138 2021-03-04.11:50 & foo1
```

Remove file with icommand
```
irm -f /nlmumc/home/rods/mounted_coll/foo1
```
List the mounted collection in iRODS
```
ils -L /nlmumc/home/rods/mounted_coll

# stdout:
/nlmumc/home/rods/mounted_coll:
```

### Observed behavior (including steps to reproduce, if applicable)
In iRODS 4.2.8 overwriting the file results in `-817000 CAT_UNKNOWN_FILE` error and a 0 byte file.
Next to that, `irm -f` is not able to delete the file.

Call the rule 
```
irule -r irods_rule_engine_plugin-irods_rule_language-instance -F overwriteFileInMountedColl.r "*irodsColl='/nlmumc/home/rods/mounted_coll'" "*irodsResc='rootResc'" "*phyDir='/tmp/mounted-coll'"
```
stdout:
```
remote addresses: 172.19.0.15 ERROR: rcExecMyRule error.  status = -817000 CAT_UNKNOWN_FILE
Level 0: DEBUG: execMicroService3: error when executing microservice
line 29, col 4
	msiDataObjCreate(*Obj2,*OFlags2,*D_FD2);
	^

caused by: DEBUG: msiDataObjCreate: rsDataObjCreate failed for /nlmumc/home/rods/mounted_coll/foo1, status = -817000

caused by: db_mod_data_obj_meta_op - Failed to find file in database by its logical path.
```

rodsLog
```
Mar  4 13:05:21 pid:1815 remote addresses: 172.19.0.15, 172.19.0.9 ERROR: [-]   /irods_git_repo/server/api/src/rsModDataObjMeta.cpp:164:int _rsModDataObjMeta(rsComm_t *, modDataObjMeta_t *) :  status [CAT_UNKNOWN_FILE]  errno [] -- message [_rsModDataObjMeta - Failed to modify the database for object "/nlmumc/home/rods/mounted_coll/foo1" - CAT_UNKNOWN_FILE ]

Mar  4 13:05:21 pid:1815 NOTICE: _rsModDataObjMeta - Failed updating the database with object info.
Mar  4 13:05:21 pid:1815 remote addresses: 172.19.0.15, 172.19.0.9 ERROR: dataOpen: Could not update size of data object [status = -817000, path = /nlmumc/home/rods/mounted_coll/foo1]
Mar  4 13:05:21 pid:1815 DEBUG: msiDataObjCreate: rsDataObjCreate failed for /nlmumc/home/rods/mounted_coll/foo1, status = -817000
Mar  4 13:05:21 pid:1815 remote addresses: 172.19.0.15, 172.19.0.9 ERROR: executeRuleAction Failed for msiDataObjCreate status = -817000 CAT_UNKNOWN_FILE
Mar  4 13:05:21 pid:1815 NOTICE: executeRuleBody: Microservice or Action msiDataObjCreate Failed with status -817000
Mar  4 13:05:21 pid:1815 NOTICE: execRuleNodeRes: applyRule Failed: IRULE_overwriteFileInMountedColl with status -817000
Mar  4 13:05:21 pid:1815 remote addresses: 172.19.0.15, 172.19.0.9 ERROR: executeRuleAction Failed for IRULE_overwriteFileInMountedColl status = -817000 CAT_UNKNOWN_FILE

```

List files in iRODS
```
ils -L /nlmumc/home/rods/mounted_coll

# stdout:
/nlmumc/home/rods/mounted_coll:
  rods         mountP rootResc;demoResc            0 2021-03-04.13:05 & foo1
```

Remove file with icommand 
```
irm -f /nlmumc/home/rods/mounted_coll/foo1

# rodsLog
Mar  4 13:09:50 pid:1838 NOTICE: Replica is not in a vault. Unregistering replica and leaving it on disk as-is [data_object=/nlmumc/home/rods/mounted_coll/foo1, physical_object=/tmp/mounted-coll/foo1].
```

List the mounted collection in iRODS
```
ils -L /nlmumc/home/rods/mounted_coll

# stdout:
/nlmumc/home/rods/mounted_coll:
      rods         mountP rootResc;demoResc            0 2021-03-04.13:05 & foo1
```

**Comment by trel** [comment]
Feels similar to https://github.com/irods/irods/issues/5426

**Comment by trel** [comment]
Notably, this issue is failing/erroring with files much smaller than the 1MB threshold discovered in #5426

**Comment by alanking** [comment]
Reproduced (different error, having to do with `replica_state_table`) on master...
```
$ irule -r irods_rule_engine_plugin-irods_rule_language-instance -F overwriteFileInMountedColl.r "*irodsColl='/tempZone/home/rods/mounted_coll'" "*irodsResc='rootResc'" "*phyDir='/tmp/mounted-coll'"
remote addresses: x.x.x.x ERROR: rcExecMyRule error.  status = -1800000 KEY_NOT_FOUND
Level 0: DEBUG: execMicroService3: error when executing microservice
line 29, col 4
    msiDataObjCreate(*Obj2,*OFlags2,*D_FD2);
    ^

caused by: DEBUG: msiDataObjCreate: rsDataObjCreate failed for /tempZone/home/rods/mounted_coll/foo1, status = -1800000


$ ils -Lr
/tempZone/home/rods:
  C- /tempZone/home/rods/mounted_coll  mountP    /tmp/mounted-coll   rootResc;leaf0
/tempZone/home/rods/mounted_coll:
  rods         mountP rootResc;leaf0            0 2021-05-11.19:24 & foo1
$ ls -al /tmp/mounted-coll/
total 2424
drwxr-x--- 2 irods irods    4096 May 11 19:24 .
drwxrwxrwt 1 root  root  2465792 May 11 19:24 ..
-rw------- 1 irods irods       0 May 11 19:24 foo1
$ irm -f /tempZone/home/rods/mounted_coll/foo1
$ ils -Lr
/tempZone/home/rods:
  C- /tempZone/home/rods/mounted_coll  mountP    /tmp/mounted-coll   rootResc;leaf0
/tempZone/home/rods/mounted_coll:
  rods         mountP rootResc;leaf0            0 2021-05-11.19:24 & foo1
```

...and fixed after a couple of tweaks...?
```
$ irule -r irods_rule_engine_plugin-irods_rule_language-instance -F overwriteFileInMountedColl.r "*irodsColl='/tempZone/home/rods/mounted_coll'" "*irodsResc='rootResc'" "*phyDir='/tmp/mounted-coll'"
Wrote file /tempZone/home/rods/mounted_coll/foo1
Overwrote file /tempZone/home/rods/mounted_coll/foo1
$ ils -Lr                     
/tempZone/home/rods:                                                 
  C- /tempZone/home/rods/mounted_coll  mountP    /tmp/mounted-coll   rootResc;leaf0
/tempZone/home/rods/mounted_coll:
  rods         mountP rootResc;leaf0          138 2021-05-11.19:19 & foo1

$ ls -l /tmp/mounted-coll/foo1
-rw------- 1 irods irods     138 May 11 19:19 foo1
$ irm -f /tempZone/home/rods/mounted_coll/foo1
$ ils -Lr
/tempZone/home/rods:
  C- /tempZone/home/rods/mounted_coll  mountP    /tmp/mounted-coll   rootResc;leaf0
$ ls -l /tmp/mounted-coll/foo1
ls: cannot access '/tmp/mounted-coll/foo1': No such file or directory
```

I will add a test that runs the rule code provided by @PaulVanSchayck

**Comment by trel** [comment]
checkboxed.

**Comment by PaulVanSchayck** [comment]
Thanks @alanking and @trel for getting this fixed. And yes it was indeed Maarten's test script, not mine. 

This issue was blocking our upgrade to 4.2.8, so it's very nice to get it fixed for 4.2.9. We realize the mounted collection are a  exotic functionality. So we're happy for you to keep it alive.

I wouldn't be surprised if we moved away from it in the next few months though. And instead use the ingest engine, and/or direct ingests over davrods to iRODS.
